lib/opam_cmd.ml

```diff
@@ -125,27 +125,89 @@ let classify_package ~package ~dev_repo ~archive ~pins () =
             | false -> (`Unknown host, tag) )
           | None -> err "dev-repo without host" ) )
 
-let check_if_dune ~root package =
-  Exec.get_opam_depends ~root (string_of_package package)
-  >>| List.exists (fun l -> l = "jbuilder" || l = "dune")
+let extract_opam_value = function
+  | None -> None
+  | Some str ->
+      let res =
+        try Ok (OpamParser.value_from_string str "")
+        with OpamLexer.Error _ ->
+          let error_message = Printf.sprintf "Failed to parse %S" str in
+          Error (`Msg error_message)
+      in
+      Some res
 
-let get_opam_info ~root ~pins package =
-  Exec.get_opam_dev_repo ~root (string_of_package package) >>= fun dev_repo ->
-  Exec.get_opam_archive_url ~root (string_of_package package) >>= fun archive ->
-  let dev_repo, tag = classify_package ~package ~dev_repo ~archive ~pins () in
-  check_if_dune ~root package >>= fun is_dune ->
-  Logs.info (fun l ->
-      l "Classified %a as %a with tag %a"
-        Fmt.(styled `Yellow pp_package)
-        package pp_repo dev_repo
-        Fmt.(option string)
-        tag );
-  let tag =
-    match List.find_opt (fun { pin; _ } -> package.name = pin) pins with
-    | Some { tag; _ } -> tag
-    | None -> tag
+let parse_string ~field ~package v =
+  let open OpamParserTypes in
+  match extract_opam_value v with
+  | Some (Ok (String (_, v))) -> Ok v
+  | Some (Ok (List (_, []))) -> Ok ""
+  | None -> Ok ""
+  | _ ->
+      R.error_msg
+        (Fmt.strf "Unable to parse opam string.\nTry `opam show --normalise -f %s: %s`" field
+           package)
+
+let parse_dev_repo = parse_string ~field:"dev-repo"
+
+let parse_archive_url ~package v =
+  parse_string ~field:"url.src" ~package v >>= function "" -> Ok None | uri -> Ok (Some uri)
+
+let parse_opam_depends ~package v =
+  let open OpamParserTypes in
+  match extract_opam_value v with
+  | Some (Ok (List (_, vs))) ->
+      let ss =
+        List.fold_left
+          (fun acc -> function Option (_, String (_, v), _) | String (_, v) -> v :: acc | _ -> acc
+            )
+          [] vs
+      in
+      Logs.debug (fun l -> l "Depends for %s: %s" package (String.concat ~sep:" " ss));
+      Ok ss
+  | None -> Ok []
+  | _ ->
+      R.error_msg
+        (Fmt.strf
+           "Unable to parse opam depends for %s\n\
+            Try `opam show --normalise -f depends: %s` manually"
+           package package)
+
+let get_opam_info ~root ~pins packages =
+  let fields = [ "name"; "dev-repo:"; "url.src:"; "depends:" ] in
+  Exec.run_opam_show ~root ~packages ~fields >>= fun lines ->
+  Opam_show_result.make lines >>= fun data ->
+  let packages =
+    List.map
+      (fun pkg ->
+        let name = pkg.name in
+        let archive_url_line = Opam_show_result.get ~package:name ~field:"url.src:" data in
```

Comment:
I suggest moving calls to `Opam_show_result.get` to `extract_opam_value`. The `package` and `field` arguments are needed in the `parse_*` functions and are passed in different ways.

Reply:
Done. Indeed it improves readability.